In [1]:
import graphlab

In [2]:
image_train = graphlab.SFrame('image_train_data/')
image_test = graphlab.SFrame('image_test_data/')
graphlab.canvas.set_target('ipynb')

[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: /tmp/graphlab_server_1537798229.log


This non-commercial license of GraphLab Create for academic use is assigned to r38411@yahoo.com and will expire on August 31, 2019.


In [3]:
image_train.head()

Columns:
	id	int
	image	Image
	label	str
	deep_features	array
	image_array	array

Rows: 10

Data:
+-----+----------------------+------------+-------------------------------+
|  id |        image         |   label    |         deep_features         |
+-----+----------------------+------------+-------------------------------+
|  24 | Height: 32 Width: 32 |    bird    | [0.242871761322, 1.0954537... |
|  33 | Height: 32 Width: 32 |    cat     | [0.525087952614, 0.0, 0.0,... |
|  36 | Height: 32 Width: 32 |    cat     | [0.566015958786, 0.0, 0.0,... |
|  70 | Height: 32 Width: 32 |    dog     | [1.12979578972, 0.0, 0.0, ... |
|  90 | Height: 32 Width: 32 |    bird    | [1.71786928177, 0.0, 0.0, ... |
|  97 | Height: 32 Width: 32 | automobile | [1.57818555832, 0.0, 0.0, ... |
| 107 | Height: 32 Width: 32 |    dog     | [0.0, 0.0, 0.220677852631,... |
| 121 | Height: 32 Width: 32 |    bird    | [0.0, 0.23753464222, 0.0, ... |
| 136 | Height: 32 Width: 32 | automobile | [0.0, 0.0, 0.0, 0.0, 0.0, ... |
| 138 | Height: 32 Width: 32 |    bird    | [0.658935725689, 0.0, 0.0,... |
+-----+----------------------+------------+-------------------------------+
+-------------------------------+
|          image_array          |
+-------------------------------+
| [73.0, 77.0, 58.0, 71.0, 6... |
| [7.0, 5.0, 8.0, 7.0, 5.0, ... |
| [169.0, 122.0, 65.0, 131.0... |
| [154.0, 179.0, 152.0, 159.... |
| [216.0, 195.0, 180.0, 201.... |
| [33.0, 44.0, 27.0, 29.0, 4... |
| [97.0, 51.0, 31.0, 104.0, ... |
| [93.0, 96.0, 88.0, 102.0, ... |
| [35.0, 59.0, 53.0, 36.0, 5... |
| [205.0, 193.0, 195.0, 200.... |
+-------------------------------+
[10 rows x 5 columns]

In [4]:
image_train['label'].sketch_summary()


+------------------+-------+----------+
|       item       | value | is exact |
+------------------+-------+----------+
|      Length      |  2005 |   Yes    |
| # Missing Values |   0   |   Yes    |
| # unique values  |   4   |    No    |
+------------------+-------+----------+

Most frequent items:
+-------+------------+-----+-----+------+
| value | automobile | cat | dog | bird |
+-------+------------+-----+-----+------+
| count |    509     | 509 | 509 | 478  |
+-------+------------+-----+-----+------+


# Question 1: there are 4 unique labels: automobile, cat, dog and bird in image_train

## Question 2: split the data into 4 groups based on the label

In [5]:
data_list = {}
data_model = {}
i_label = image_train['label'].unique()
for i in range(len(i_label)):
    data_list[i_label[i]] = image_train.filter_by(i_label[i], 'label')
    data_model[i_label[i]] = graphlab.nearest_neighbors.create(data_list[i_label[i]],
                                                                features = ['deep_features'],
                                                                label = 'id')

Starting brute force nearest neighbors model training.

Starting brute force nearest neighbors model training.

Starting brute force nearest neighbors model training.

Starting brute force nearest neighbors model training.

In [6]:
the_cat = image_test[0:1]
the_cat['image'].show()

In [7]:
q2_nearest = {}
q2_nearest_1 = {}
for i in range(len(i_label)):
    q2_nearest[i_label[i]] = data_model[i_label[i]].query(the_cat)
    q2_nearest_1[i_label[i]] = q2_nearest[i_label[i]][0]['reference_label']
q2_nearest_1

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.209205    | 10.913ms     |

| Done         |         | 100         | 52.265ms     |

+--------------+---------+-------------+--------------+

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 11.434ms     |

| Done         |         | 100         | 63.265ms     |

+--------------+---------+-------------+--------------+

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 12.253ms     |

| Done         |         | 100         | 87.963ms     |

+--------------+---------+-------------+--------------+

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 11.135ms     |

| Done         |         | 100         | 66.671ms     |

+--------------+---------+-------------+--------------+

{'automobile': 49314, 'bird': 49573, 'cat': 16289, 'dog': 16976}

In [9]:
image_train[image_train['id']==16976].show()

# Question 2 answer: nearest neighbor of the the_cat: 
# automobile model: 49314, bird_model: 49573, cat_model: 16289, dog_model: 16976

## Question 3

In [8]:
q2_nearest['automobile']

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      49314      | 39.6710582792 |  1   |
|      0      |      12389      | 40.2107393099 |  2   |
|      0      |      15299      | 41.2103305258 |  3   |
|      0      |      40118      | 41.3292164653 |  4   |
|      0      |      17111      |  41.352687775 |  5   |
+-------------+-----------------+---------------+------+
[5 rows x 4 columns]

In [9]:
for i in q2_nearest.keys():
    print i, ': ', 'avg distance = ', sum(q2_nearest[i]['distance'])/len(q2_nearest[i]['distance'])

automobile :  avg distance =  40.754806471
cat :  avg distance =  36.1557307098
bird :  avg distance =  39.2939524591
dog :  avg distance =  37.7707113618


# Question 3: avg distance from each model: 
# automobile: 40.7548, cat: 36.155, bird: 39.2939, dog: 37.77

## Question 4

In [10]:
## split the image test data based on the label

In [11]:
test_data_group = {}
for i in range(len(i_label)):
    test_data_group[i_label[i]] = image_test.filter_by(i_label[i], 'label')

In [12]:
def test_group(which_test_group):
    neighbors = {}
    for i in range(len(i_label)):
        neighbors[i_label[i]] = data_model[i_label[i]].query(test_data_group[which_test_group], k=1)
        print i_label[i]
        print neighbors[i_label[i]].head()
    return neighbors

In [13]:
def build_distance(which_test_group):
    group_distance = graphlab.SFrame()
    for i in range(len(i_label)):
        col_name = which_test_group + '-' + i_label[i]
        group_distance[col_name] = neighbors[i_label[i]]['distance']
    return group_distance

In [14]:
def is_group_correct(row):
    same_group_name = test_group_name + '-' + test_group_name
    if min(row, key=row.get) == same_group_name:
        return 1
    else:
        return 0
#cat_distance.apply(is_cat_correct).sum()

In [15]:
test_group_name = 'dog'
neighbors = test_group(test_group_name)
group_distance = build_distance(test_group_name)
ans = group_distance.apply(is_group_correct).sum()
print "test image group = " + test_group_name + ":" + "matches " + str(ans) + " in the same image train group"

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 120000  | 25.1046     | 300.576ms    |

| Done         | 478000  | 100         | 323.398ms    |

+--------------+---------+-------------+--------------+

bird
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      44658      | 41.7538647304 |  1   |
|      1      |       9215      | 41.3382958925 |  1   |
|      2      |      36675      | 38.6157590853 |  1   |
|      3      |      12582      | 37.0892269954 |  1   |
|      4      |      36122      |  38.272288694 |  1   |
|      5      |       8736      | 39.1462089236 |  1   |
|      6      |      38991      |  40.523040106 |  1   |
|      7      |      44177      | 38.1947918393 |  1   |
|      8      |       4549      | 40.1567131661 |  1   |
|      9      |      40225      | 45.5597962603 |  1   |
+-------------+-----------------+---------------+------+
[10 rows x 4 columns]



Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 277.336ms    |

| Done         | 509000  | 100         | 306.781ms    |

+--------------+---------+-------------+--------------+

dog
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      49803      | 33.4773590373 |  1   |
|      1      |       5755      | 32.8458495684 |  1   |
|      2      |      20715      | 35.0397073189 |  1   |
|      3      |      13387      | 33.9010327697 |  1   |
|      4      |      12089      | 37.4849250909 |  1   |
|      5      |       6094      |  34.945165344 |  1   |
|      6      |       3431      | 39.0957278345 |  1   |
|      7      |       6184      | 37.7696131032 |  1   |
|      8      |       2167      | 35.1089144603 |  1   |
|      9      |       7776      | 43.2422832585 |  1   |
+-------------+-----------------+---------------+------+
[10 rows x 4 columns]



Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 269.533ms    |

| Done         | 509000  | 100         | 297.448ms    |

+--------------+---------+-------------+--------------+

cat
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |        33       | 36.4196077068 |  1   |
|      1      |      30606      | 38.8353268874 |  1   |
|      2      |       5545      | 36.9763410854 |  1   |
|      3      |      19631      | 34.5750072914 |  1   |
|      4      |       7493      |  34.778824791 |  1   |
|      5      |      47044      | 35.1171578292 |  1   |
|      6      |      13918      | 40.6095830913 |  1   |
|      7      |      10981      | 39.9036867306 |  1   |
|      8      |      45456      | 38.0674700168 |  1   |
|      9      |      44673      | 42.7258732951 |  1   |
+-------------+-----------------+---------------+------+
[10 rows x 4 columns]



Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 309.077ms    |

| Done         | 509000  | 100         | 338.811ms    |

+--------------+---------+-------------+--------------+

automobile
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      33859      | 41.9579761457 |  1   |
|      1      |       2046      | 46.0021331807 |  1   |
|      2      |      19594      | 42.9462290692 |  1   |
|      3      |      11000      | 41.6866060048 |  1   |
|      4      |      19594      | 39.2269664935 |  1   |
|      5      |      49314      | 40.5845117698 |  1   |
|      6      |      40822      | 45.1067352961 |  1   |
|      7      |      44997      | 41.3221140974 |  1   |
|      8      |      33859      | 41.8244654995 |  1   |
|      9      |      33859      | 45.4976929401 |  1   |
+-------------+-----------------+---------------+------+
[10 rows x 4 columns]

test image group = dog:matches 678 in the same image train group


In [16]:
cat_neighbors = {}
test_group_name = 'cat'
for i in range(len(i_label)):
    cat_neighbors[i_label[i]] = data_model[i_label[i]].query(test_data_group[test_group_name], k=1)
    print i_label[i]
    print cat_neighbors[i_label[i]].head()

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 120000  | 25.1046     | 273.096ms    |

| Done         | 478000  | 100         | 311.112ms    |

+--------------+---------+-------------+--------------+

bird
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      49573      |  38.074265869 |  1   |
|      1      |       4242      | 36.3674024138 |  1   |
|      2      |      44658      | 35.3039394947 |  1   |
|      3      |      12533      | 38.8944029601 |  1   |
|      4      |       6944      | 34.2820409875 |  1   |
|      5      |      37315      | 44.5352170178 |  1   |
|      6      |      15910      | 34.0290595084 |  1   |
|      7      |      49573      | 39.0236924983 |  1   |
|      8      |       3517      | 40.8334054297 |  1   |
|      9      |      19224      | 40.1258835601 |  1   |
+-------------+-----------------+---------------+------+
[10 rows x 4 columns]



Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 299.701ms    |

| Done         | 509000  | 100         | 317.377ms    |

+--------------+---------+-------------+--------------+

dog
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      16976      | 37.4642628784 |  1   |
|      1      |      40938      | 29.3472319585 |  1   |
|      2      |      15549      | 32.2599640475 |  1   |
|      3      |      35905      | 35.3852085188 |  1   |
|      4      |       5755      | 30.0442985088 |  1   |
|      5      |      42054      | 35.4741000424 |  1   |
|      6      |      24663      | 32.5845275226 |  1   |
|      7      |      16560      | 37.6502852614 |  1   |
|      8      |      19683      | 36.9801353512 |  1   |
|      9      |       5755      | 41.1259410707 |  1   |
+-------------+-----------------+---------------+------+
[10 rows x 4 columns]



Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 280.291ms    |

| Done         | 509000  | 100         | 302.427ms    |

+--------------+---------+-------------+--------------+

cat
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      16289      |  34.623719208 |  1   |
|      1      |      13094      | 33.8680579302 |  1   |
|      2      |      18188      | 32.4615168902 |  1   |
|      3      |       9846      | 35.7708210254 |  1   |
|      4      |      16904      | 31.1577686417 |  1   |
|      5      |      16018      | 41.3986035847 |  1   |
|      6      |      29793      | 30.9894594959 |  1   |
|      7      |      39130      | 37.0814607387 |  1   |
|      8      |      42646      | 39.9883863688 |  1   |
|      9      |      39777      | 39.7076633097 |  1   |
+-------------+-----------------+---------------+------+
[10 rows x 4 columns]



Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 287.875ms    |

| Done         | 509000  | 100         | 314.343ms    |

+--------------+---------+-------------+--------------+

automobile
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      49314      | 39.6710582792 |  1   |
|      1      |      46858      | 43.0089056688 |  1   |
|      2      |      19594      | 38.6010006604 |  1   |
|      3      |       302       | 39.3566307091 |  1   |
|      4      |      20750      | 38.3572372618 |  1   |
|      5      |       9118      | 42.0904793181 |  1   |
|      6      |      44997      | 39.0520251253 |  1   |
|      7      |       302       | 39.3058645069 |  1   |
|      8      |      46858      | 43.0248129799 |  1   |
|      9      |      20750      | 45.6749176426 |  1   |
+-------------+-----------------+---------------+------+
[10 rows x 4 columns]



In [17]:
## build the SFrame

In [18]:
cat_distance = graphlab.SFrame()
for i in range(len(i_label)):
    col_name = test_group_name + '-' + i_label[i]
    cat_distance[col_name] = cat_neighbors[i_label[i]]['distance']
cat_distance.head()

Columns:
	cat-bird	float
	cat-dog	float
	cat-cat	float
	cat-automobile	float

Rows: 10

Data:
+---------------+---------------+---------------+----------------+
|    cat-bird   |    cat-dog    |    cat-cat    | cat-automobile |
+---------------+---------------+---------------+----------------+
|  38.074265869 | 37.4642628784 |  34.623719208 | 39.6710582792  |
| 36.3674024138 | 29.3472319585 | 33.8680579302 | 43.0089056688  |
| 35.3039394947 | 32.2599640475 | 32.4615168902 | 38.6010006604  |
| 38.8944029601 | 35.3852085188 | 35.7708210254 | 39.3566307091  |
| 34.2820409875 | 30.0442985088 | 31.1577686417 | 38.3572372618  |
| 44.5352170178 | 35.4741000424 | 41.3986035847 | 42.0904793181  |
| 34.0290595084 | 32.5845275226 | 30.9894594959 | 39.0520251253  |
| 39.0236924983 | 37.6502852614 | 37.0814607387 | 39.3058645069  |
| 40.8334054297 | 36.9801353512 | 39.9883863688 | 43.0248129799  |
| 40.1258835601 | 41.1259410707 | 39.7076633097 | 45.6749176426  |
+---------------+---------------+---------------+----------------+
[10 rows x 4 columns]

In [19]:
## now compute how many of them dog-dog is min distance

In [20]:
def is_cat_correct(row):
    if min(row, key=row.get) == "cat-cat":
        return 1
    else:
        return 0
cat_distance.apply(is_cat_correct).sum()

548